### ISSUE TITLE
Proposal: Support for New Processing Mode in Ext-proc Filter for Large Payloads Processing

### ISSUE THREAD

**Issue by xr** [envoyproxy/envoy#32090]
I want to propose adding supporting for a new processing mode in ext-proc (sth like STREAM_TO_BUFFER), this mode would enable streaming the request body to an external server in sync mode, where the side stream server would buffer the whole payload then stream back to envoy in chunks.

## Motivation
1. Keep the iteration stopped while streaming the chunks to the side stream server so that side stream server could cover the case that required buffering/processing the whole payload externally before determining whether the request should proceed further in the filter chain or not.
2. Provide support for 1:N mapping or N:1 mapping between the request and response message ratio to enhance flexibility. This capability can address use cases, such as enabling the side stream server to respond in chunks.

## Expected Behavior for the new mode

1. The ext-proc filter sends chunks to the external server as soon as they arrive.
3. Corresponding responses for each chunk are expected from the external server to confirm receipt.
4. Even if a response is received for each chunk, the ext-proc keep stoping iteration in the filter chain while continue streaming all the chunks to external server.
5. The external server will buffer the whole payload and process it externally.
6. After processing the payload, the server starts sending back the processed payload in chunks.
7. The ext-proc is expected to receive in stream mode and immediately inject each chunks into the filter chain.

## What we have tried so far

We currently utilize the **"STREAM"** mode to send chunks; however, it fails to meet our use case due to the following reasons:

1. The "STREAM" mode chunks operates on a 1:1 mapping, where each chunk sent to the external server corresponds to a chunk sent back. The ext_proc injects each chunk immediately into the filter chain upon receipt, **_preventing it from blocking while streaming all the chunks to the external server._**

2. Attempting to address the issue, we experimented with responding using the `clear_body` mutation for early chunks and only response the whole payload back in the last chunk, intending to avoid upstream servers from receiving early chunks as ext_proc is not in blocking mode while streaming chunks. Unfortunately, this approach presented challenges. The ext_proc will eventually receive the entire body in the last single message:
    - This will cause gRPC size limitation, for example, when using the Google gRPC client, there are limitations on the maximum message size for gRPC responses. 
    - Even without these limitations, ext_proc replaces the last buffered chunk with the entire response body and injects them in a single batch. This behavior results in buffering the entire payload in Envoy again. like here: https://github.com/envoyproxy/envoy/blob/5d2f3ae3daf78b9e4001993f25d0d7a6082f04f0/source/extensions/filters/http/ext_proc/processor_state.cc#L271

Hope this highlights the reason behind the proposal, and we believe that the new mode could further empower the ext_proc ability to cover such use case to process large payloads externally, thanks!

**Comment by jmarantz** [comment]
I'd like to generalize this slightly to make sure we consider a similar use-case. The journey described above suggests the sidestream server will want to buffer the entire response before responding to any of it. This makes sense to me for the use-case of image processing. You might have a 10 megabyte image that you want to stream into the sidestream server, which requires the entire image in memory before it processes it, reducing it to 100k (say a lossy PNG to WEBP transformation). It may then stream the data back, and the processed data may have a much different size.

Another use-case is non-lossy compression (gzip, brotli). In this case the sidestream server may choose settings that involve buffering some amount of data, say 64k. As the data streams in (say 4k at a time), once 64k is collected in the sidestream server, it runs this through a streaming gzip optimizer and streams the data back, also 4k at a time. This may not correspond to the chunks that get streamed in because gzip may vary in compression ratio based on the data.

The difference is that in the case described above, the data is fully collected before it streams back. In the lossless gzip example the data streams in both directions concurrently.

**Comment by yanjunxiang-google** [comment]
It looks like case 1) is a special case for case 2), which support side stream server buffering random amount of data before sending response, which may or may not contain end_of_stream = true.

**Comment by jmarantz** [comment]
Agreed -- I think the original description stated the requirement but then gave a more constrained example, and I just wanted a more general example.

**Comment by yanjunxiang-google** [comment]
Adding @htuch @yanavlasov @gbrail

**Comment by yanjunxiang-google** [comment]
I am thinking adding a new proto field in the BodyMutation message as below can resolve the issue with minimum ext_proc filter state machine change:

https://github.com/envoyproxy/envoy/blob/5c821febbbba530315295a3c3dd3481364f89890/api/envoy/service/ext_proc/v3/external_processor.proto#L366

message BodyMutation {
  oneof mutation {
    // The entire body to replace
    bytes body = 1;

    // Clear the corresponding body chunk
    bool clear_body = 2;
  }

// In STREAMED mode, sometimes, when side stream server receives a body chunk, call it received_chunk, it may send multiple chunks back to Envoy.
//  This *more_chunks* field tells Envoy whether there will be more body response back for that received_chunk.
// When Envoy receives this response message, if * more_chunks is false, which is the default case,
// it does normal body mutation processing. When it is true, which indicates there are more pieces for this received_chunk
// coming, Envoy ext_proc filter will not pop up that chunk from the queue. It just inject the body in the response to the filter chain.
// The side stream server must set *multiple_chunks* to false when the last piece for this chunk is responded.  And with that,
//  Envoy ext_proc filter will now pop up that received_chunk from the queue, and inject the body in response to the filter chain

+  bool more_chunks
}

**Comment by yanjunxiang-google** [comment]
So, for case 1), i.e, @xr 's original issue,  the side stream server sends a "clear_body" response for each chunk received (not for the last_chunk), with "more_chunks" set to false.  This will help disable the ext_proc request timer. Then when the side stream server receive the last_chunk, it process it, then send multiple chunks back, with the first a few chunks with "body" contains data, and "more_chunks" to true. The last response chunk with "body" contains data, and more_chunks to false.

**Comment by yanjunxiang-google** [comment]
For case 2),  it follows the same logic as above, except that the side_stream server start sending response with "body" contains data not for the last_chunk, but any time when it deems it received enough data for response.

**Comment by yanjunxiang-google** [comment]
Adding more ext_proc owners @tyxia @stevenzzzz

**Comment by stevenzzzz** [comment]
I'd ping even more folks here. :)

summon @rshriram

**Comment by htuch** [comment]
I think I'm still a bit stuck on motivation:

> The "STREAM" mode chunks operates on a 1:1 mapping, where each chunk sent to the external server corresponds to a chunk sent back. The ext_proc injects each chunk immediately into the filter chain upon receipt, preventing it from blocking while streaming all the chunks to the external server.

Why is this an issue for Envoy OOM? I think I'm missing something there, because, you shouldn't have to block the pipeline while streaming chunks to the ext_proc server if you've received a response for a chunk, that bit seems WAI?

**Comment by tyxia** [comment]
+1 I also think that OOM is probably not motivation here. Please note, in STREAMED mode, we still do buffering in ext_proc while streaming out the data. That is exactly how to prevent OOM on that path. So if your downstream payload is too large or it is too fast, it will be flow-controlled rather than being offload i think.  Otherwise, I feel such free streaming actually will cause OOM. (e.g., OOM during offload.)

I see the motivation here is actually more about a new mode that does not have strict 1:1 mapping restriction .  But still, i think for case like payload/mutation is too large or processing rate difference causing data backup etc, we should flow-control to prevent OOM.

**Comment by xr** [comment]
> So, for case 1), i.e, @xr 's original issue, the side stream server sends a "clear_body" response for each chunk received (not for the last_chunk), with "more_chunks" set to false. This "clear_body" response will reduce the ext_proc queue, also disable the ext_proc request timer. Then when the side stream server receive the last_chunk, it process it, then send multiple chunks back, with the first a few chunks with "body" contains actual data, and "more_chunks" to true. When Envoy receives such response, it will inject the data to filter chain. The side stream server sends the last response chunk with "body" contains data, and more_chunks to false. Envoy will inject the data to filter chain then continue the filter iteration.

@yanjunxiang-google thanks! this looks promising and it will allow 1:N (request/response) mapping, hope we could move forward with this if possible.

**Comment by xr** [comment]
> I think I'm still a bit stuck on motivation:
> 
> > The "STREAM" mode chunks operates on a 1:1 mapping, where each chunk sent to the external server corresponds to a chunk sent back. The ext_proc injects each chunk immediately into the filter chain upon receipt, preventing it from blocking while streaming all the chunks to the external server.
> 
> Why is this an issue for Envoy OOM? I think I'm missing something there, because, you shouldn't have to block the pipeline while streaming chunks to the ext_proc server if you've received a response for a chunk, that bit seems WAI?

The use case is that:
1. our side stream server need to buffer the whole payload before processing it and decide whether or not the request is allowed to move forward to the upstream or not, this is why we want to block the pipeline while streaming.
2. we don't want to use the BUFFER MODE and let envoy to buffer all the payload for us, as the payload could be large + high throughput, so we prefer to offloading the buffering/processing work externally.

**Comment by xr** [comment]
> +1 I also think that OOM is probably not motivation here. Please note, in STREAMED mode, we still do buffering in ext_proc while streaming out the data. We need it for applying body mutation and it also help to prevent OOM on that path. So if your downstream payload is too large or it is too fast, it will be flow-controlled rather than being offload i think. Otherwise, I feel such free streaming actually will cause OOM. (e.g., OOM during offload.)
> 
> I see the motivation here is actually more about a new mode that does not have strict 1:1 mapping restriction, which i agree it make sense for some use cases mentioned [here](https://github.com/envoyproxy/envoy/issues/32090#issuecomment-1915347779). But still, i think for case like payload/mutation is too large or processing rate difference causing data backup etc, we should flow-control to prevent OOM.

Thanks, good points, a new mode does not have strict 1:1 mapping is clearly a motivation, i'll edit to my original post

**Comment by tyxia** [comment]
Furthermore, I think the mode actually proposed here is a complete streaming mode with no Envoy buffering during the callout to sidecar server, @jmarantz is my understanding correct?  

If so, then it can not be achieved by justing adding proto field, as both STREAM and BUFFER mode are still buffering. Actually there is probably no need to buffer in this case as it is not 1:1 mutating the mainstream data.

It is a bit similar to async mode. I think robust side stream flow control is the first thing to make such mode possible.  We also need to evaluate the state machine. Besides, we need to be very careful about vulnerability, reliability aspects that could be exploited by sidecar server in this mode.

**Comment by yanjunxiang-google** [comment]
The issue is not OOM. The issue is that if the side stream server send the entire response body in one shot,  it will over gRPC channel limits. Thus the side stream server has to split the response body into smaller chunks and stream them back. So, it is 1:N case.

**Comment by jmarantz** [comment]
RE "no envoy buffering" -- I think that can be implementation dependent. Like if the filter-chain provides a bunch of 1 byte messages maybe envoy might want to buffer up to (say) 64k before sending to ext_proc, idk. But it should have bounded buffering.

**Comment by yanjunxiang-google** [comment]
"Like if the filter-chain provides a bunch of 1 byte messages maybe envoy might want to buffer up to (say) 64k before sending to ext_proc",  yeah this is a good point. We had discussed this before. We do not have this kind of buffering at this moment. We plan to add that. But that's a separate issue.

**Comment by xr** [comment]
Thanks for the comments! @tyxia @jmarantz @yanjunxiang-google 

Actually, the proposal is not intended for a "complete stream" mode. I believe that a buffering mechanism is necessary. However, the chunk/buffer size could be optional for configuration, allowing for fine-tuning on a case-by-case basis.

In this sense, the current STREAM mode is good; however, the missing part is the 1:N/N:1 request/response ratio, providing more flexibility for the side stream server beyond current support:

- Allowing it to buffer the entire payload while stopping iteration to cover those use cases.
- Streaming back in chunks to avoid one-shot big chunk back.

In fact, the intended behavior of the proposal could be implemented based on the current STREAM mode with a new field `more_chunks` as @yanjunxiang-google suggested [above](https://github.com/envoyproxy/envoy/issues/32090#issuecomment-1915742972). It would be great if we could move forward with this solution.

**Comment by jmarantz** [comment]
I suppose the chunk queue (which I had not been aware of till now) is needed if we can't back-pressure in any way. In an HTTP context I think back-pressuring is done by readDisable and I realize this is complex to get right. In a gRPC context I thought there could be flow-control protocol of some sort to avoid the need for unbounded buffering (and the memory pressure resulting from that).

I'm not sure exactly what's needed from an implementation perspective. I am just trying to understand a path where it's at least possible for an implementation to use bounded memory in the context of a sidestream that may greatly change the structure of the response chunks.

**Comment by yanjunxiang-google** [comment]
The chunk queue is used for the following cases:
1) flow control in the fast downstream and slow side stream scenario.
2) When response coming back, knows which data to mutate, for example, if response is "clear_body", the chunk queue provides the corresponding data to be cleared.
3)  In future the 1:N case, it is needed to map that "N" to "1".

**Comment by yanjunxiang-google** [comment]
@xr we are prototyping and will send out a PR once it is ready.

**Comment by tyxia** [comment]
How about N:1 or even random M:N case? Basically there is no determined mapping between the chunk_queue and response from external processor. 

From design perspective,  side-stream response now is not targeted to any particular 1 chunk or X chunks now.

From implementation perspective, ext_proc should/can not decide how many chunks corresponding to 1 response. We probably start from 1:1  then it could leads to that we need to handle:
- no more sidecar response  + many chunks left in the queue : You need to discard those chunks you stored. 
- more sidecar response + no chunks left in the queue: you need to allocate buffer to inject the data.

chunk queue currently is must to have for flow control which could be improved in the future.

**Comment by xr** [comment]
Regardless of the purpose for flow control of chunk_queue, if I only consider the need for handling body_mutation based on the response from the side stream server in a request/response 1:N scenario, will it still buffer the response "N" somewhere with the help of `more_chunks`, then dequeue the request 1 chunk once all N are received then move forward?  

I was thinking Envoy would stream immediately down to the filter chain as soon as received every single N response chunks. In this scenario, the side stream server could handle the mutation by itself and simply stream back whatever mutated payload, potentially eliminating the need for Envoy to perform the mutation? 

In this context, and without knowledge of other constraints, could the chunk queue be maintained only for flow control purposes, with the option of discarding the chunks after calling out to the side stream server, as they are not required for mutation thereafter? Also no need to concern the mapping afterward? @yanjunxiang-google @tyxia

**Comment by tyxia** [comment]
> I was thinking Envoy would stream immediately down to the filter chain as soon as received every single N response chunks. In this scenario, the side stream server could handle the mutation by itself and simply stream back whatever mutated payload, potentially eliminating the need for Envoy to perform the mutation?

This is exactly why I understand this proposal as complete streaming mode :) Technically sidestream responses are independent of downstream bodies in this mode and chunk_queue is redundant from this perspective.  
Note, we should still have buffering for small downstream data before callout as @jmarantz suggested for performance optimization.

I just want to clarify:  I just feel it is not 100% ideal, but if there is need to kick off this mode right away, I am not against using chunk queue since we don't have complete sidestream flow control yet. Though I think we need to handle some cases i mentioned [here](https://github.com/envoyproxy/envoy/issues/32090#issuecomment-1919585483)

**Comment by gbrail** [comment]
If I recall, we needed the chunk queue because there wasn't a guaranteed way to stop Envoy from sending the ext_proc filter a new chunk while we waited for the current chunk to be processed by the remote processor -- and even if there was, the performance impact would be huge without some amount of a "sliding window" kind of a protocol. If I recall, ext _proc tries not to tell Envoy to initiate flow control until some memory limits are reached so that we can be transmitting new chunks while still waiting for existing chunks to be processed -- that reduces latency a ton.

I think what you're proposing is indeed a "total streaming" mode, in which the ext_proc processor totally replaces the stream coming from the upstream/downstream with its own stream.

In other words, today the algorithm is something like this, for each chunk:

1. Receive chunk from filter stream
2. Save chunk on chunk queue and transmit to processor over gRPC stream
3. When matching response comes back from processor, find the right chunk that we saved on the chunk queue
4. Depending on the instructions from the processor, we may clear the chunk and turn it into an empty chunk, replace the contents of the chunk, or leave it alone
5. Let the chunk continue processing in the filter stream

And all of steps 2-5 are repeated for every chunk and can be interleaved for individual chunks.

That supports all streaming use cases including the one that you describe, as long as a single chunk is never larger than all the various buffer limits, including gRPC payload size.

I think you're proposing something more like:

1. Receive chunk from filter stream
2. Send it over gRPC stream to external processor
3. Throw the chunk away

and then in parallel, inject every chunk from the processor back in to the filter stream as soon as you get it.

This would not support all use cases, but would be better for processors that will always replace the complete stream.

I don't know how to make the Envoy filter stream do this, but I'm sure it's possible.

This leads to another problem I had, and one that I think will eventually impact you all if you start working with trailers -- there is no way in Envoy to distinguish between "end of stream" and "end of body, but there are trailers coming." I think that this will bite you if you are trying to implement this algorithm and also stick a trailer on the end -- for example, if you're trying to transform a JSON request to gRPC, or vice versa, or transform a gRPC message...

**Comment by gbrail** [comment]
And having made that long comment, my proposal would not help with flow control -- if the proxy sent chunks to the processor faster than the processor can handle, I'm not sure what happens -- there would be no flow control unless we add it to the gRPC protocol, and that would complicate things.

The original proposal, which would change step "4" above to allow the processor to send multiple chunks back instead of a single chunk, might work better.

**Comment by jmarantz** [comment]
By "gRPC protocol" I assume you mean the way in which we use gRPC between sidestream server and ext_proc, and not the implementation of gRPC itself. Is that right? I am guessing this would impose an extra round-trip per chunk, yeah? I haven't considered the details that much yet, but it seems workable, no?

I think fundamental to my thinking here is the ability for the proxy to stream through a sidestream response that's larger than available memory, at least as an option. In this case of course the sidestream would need to commit to passing through all content even if it is not going to change any of it. Obviously this mode needs to be chunked.

Clearly also a separate mode is called for with (a) bounded sizes we are comfortable buffering in memory (b) sidestream able to opt out of modifying the content so ext_proc can just pass it to the next filter in this case. Whether this mode requires chunking or not is unclear to me.

**Comment by htuch** [comment]
Based on what I'm seeing with ext_proc use cases, there are a variety of streaming scenarios, e.g. ones in which there are predominantly no mutations (e.g. DLP scanning), ones in which every chunk is mutated (e.g. compression). There are also streaming optimizations like short circuiting to only observe and rewrite some stream prefix.

> and then in parallel, inject every chunk from the processor back in to the filter stream as soon as you get it.

This would need to come with the strict contract that ordering is preserved, to avoid needing to queue / reorder on the receiving side.

>  if the proxy sent chunks to the processor faster than the processor can handle, I'm not sure what happens -- there would be no flow control unless we add it to the gRPC protocol, and that would complicate things.

I _think_ the original motivation isn't driven by protecting the processor. We also have Envoy-side flow control (watermarks) to push back if queue size grows too large.

I agree with the above observations that this is a new parallel streaming mode. I _think_ it's simpler to implement than the existing on since there is no queueing, and seems to use cases that sit somewhere between BUFFERED And STREAMED; you want the external processor to be able to fully buffer the body (presumably where there is lots of memory) and want to tightly bound the memory use on the Envoy side (where there is presumably more scarce memory).

**Comment by yanjunxiang-google** [comment]
I created this PR: https://github.com/envoyproxy/envoy/pull/32171 to address the original issue, which is to enable the side stream server be able to streaming back multiple chunks for any individual received chunk. The change to support this actually is pretty small.

I don't think we want to remove the Envoy ext_proc chunk_queue at this moment.

**Comment by xr** [comment]
Hello, i noticed the draft PR was close, I'm wondering any chances if we could move forward with it? Thank you!

**Comment by yanjunxiang-google** [comment]
This work is left aside due to some other internal priorities. We are going to resume this work soon.

**Comment by yanjunxiang-google** [comment]
I created https://github.com/envoyproxy/envoy/pull/34942 to wrap up this.  I will add a few more tests then will turn on it for review.
@xr ,  could you please patch the PR and test in your case.  I listed the steps about how to make this work for MxN streaming.

**Comment by xr** [comment]
@yanjunxiang-google thanks, will try and get back to you soon!

**Comment by yanjunxiang-google** [comment]
@xr Please let us know the result. Thanks!

**Comment by xr** [comment]
@yanjunxiang-google sorry, was busy and on vacation, taking a look now, any standard/quick way to sync your proto changes to https://github.com/envoyproxy/go-control-plane/tree/main/envoy/service/ext_proc/v3 as our ext-proc are building on top it?

**Comment by xr** [comment]
@yanjunxiang-google nvm, I compiled the new proto by myself and tested it with the patch in your PR. It looks promising and fits our use cases perfectly! Thanks for the contribution, and I hope this PR can land in the release soon! 🚀

here is the commit I used for building the envoy:
```
/mnt/envoy  version: 12bf1cb1e7311d88fd6eaa56e04d3a18944c1017/1.31.0-dev/Modified/RELEASE/BoringSSL
```

here are the concrete steps I tried to e2e test:

1. when receiving M chunks from envoy, respond  (1 ~ M-1) -th with

```
resp := &extProcProto.ProcessingResponse{
	Response: &extProcProto.ProcessingResponse_RequestBody{
		RequestBody: &extProcProto.BodyResponse{
			Response: &extProcProto.CommonResponse{
				BodyMutation: &extProcProto.BodyMutation{
					Mutation: &extProcProto.BodyMutation_ClearBody{
						ClearBody: true,
					},
					MoreChunks: false,
				},
			},
		},
	},
}

```

2. when received the last M chunk, start responding in multiple responses from side server:

```
resp := &extProcProto.ProcessingResponse{
	Response: &extProcProto.ProcessingResponse_RequestBody{
		RequestBody: &extProcProto.BodyResponse{
			Response: &extProcProto.CommonResponse{
				BodyMutation: &extProcProto.BodyMutation{
					Mutation: &extProcProto.BodyMutation_Body{
						Body: []byte("AAAAAA"),
					},
					MoreChunks: true,
				},
			},
		},
	},
}

resp2 := &extProcProto.ProcessingResponse{
	Response: &extProcProto.ProcessingResponse_RequestBody{
		RequestBody: &extProcProto.BodyResponse{
			Response: &extProcProto.CommonResponse{
				BodyMutation: &extProcProto.BodyMutation{
					Mutation: &extProcProto.BodyMutation_Body{
						Body: []byte("BBBBB"),
					},
					MoreChunks: true,
				},
			},
		},
	},
}

resp3 := &extProcProto.ProcessingResponse{
	Response: &extProcProto.ProcessingResponse_RequestBody{
		RequestBody: &extProcProto.BodyResponse{
			Response: &extProcProto.CommonResponse{
				BodyMutation: &extProcProto.BodyMutation{
					Mutation: &extProcProto.BodyMutation_Body{
						Body: []byte("CCCCCC"),
					},
					MoreChunks: false,
				},
			},
		},
	},
}
```

Then 🎉, from the upstream service I got multiple chunks:

```
request headers.
request body:
Received chunk: 6 AAAAAA
Received chunk: 5 BBBBB
Received chunk: 6 CCCCCC
Complete body received! total length: 17
```

Logs from envoy:
```
[2024-07-26 07:21:15.535][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:339] Opening gRPC stream to external processor
[2024-07-26 07:21:15.538][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.538][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:434] Sending headers message
[2024-07-26 07:21:15.538][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:443] Watermark raised on decoding
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request headers response
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:87] applying headers response. body mode = STREAMED
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:159] Sending first chunk using buffered data (45482)
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 45482 bytes, end_stream false
[2024-07-26 07:21:15.544][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.545][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:451] Watermark lowered on decoding
[2024-07-26 07:21:15.545][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:386] Continuing processing
[2024-07-26 07:21:15.546][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 183894 bytes, end_stream false
[2024-07-26 07:21:15.546][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.547][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 31856 bytes, end_stream false
[2024-07-26 07:21:15.547][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.548][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 32768 bytes, end_stream false
[2024-07-26 07:21:15.548][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.551][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 105008 bytes, end_stream false
[2024-07-26 07:21:15.551][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.555][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 649568 bytes, end_stream false
[2024-07-26 07:21:15.555][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.555][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 0 bytes, end_stream true
[2024-07-26 07:21:15.556][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:15.647][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:15.647][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:15.647][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 45482
[2024-07-26 07:21:15.647][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:15.753][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:15.753][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:15.753][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 183894
[2024-07-26 07:21:15.753][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:15.854][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:15.854][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:15.854][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 31856
[2024-07-26 07:21:15.854][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:15.954][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:15.954][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:15.954][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 32768
[2024-07-26 07:21:15.955][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:16.056][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:16.056][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:16.056][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 105008
[2024-07-26 07:21:16.056][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:16.162][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:16.162][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:16.162][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 649568
[2024-07-26 07:21:16.162][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:21.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:21.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:21.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:270] Streamed mode body response has more_chunks set to true. enableMoreChunks true
[2024-07-26 07:21:21.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:21.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:26.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:26.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:26.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:270] Streamed mode body response has more_chunks set to true. enableMoreChunks true
[2024-07-26 07:21:26.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:26.264][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction INBOUND: 30000 ms timer enabled
[2024-07-26 07:21:31.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received request body response
[2024-07-26 07:21:31.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:31.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:397] Applying body response to chunk of data. Size = 0
[2024-07-26 07:21:31.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:31.265][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:386] Continuing processing
[2024-07-26 07:21:31.266][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction OUTBOUND: 30000 ms timer enabled
[2024-07-26 07:21:31.267][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:434] Sending headers message
[2024-07-26 07:21:31.267][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:509] Watermark raised on encoding
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received response headers response
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:87] applying headers response. body mode = STREAMED
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction OUTBOUND: timer disabled
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:863] Sending a body chunk of 17 bytes, end_stream true
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:29] Traffic direction OUTBOUND: 30000 ms timer enabled
[2024-07-26 07:21:31.268][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:517] Watermark lowered on decoding
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:1075] Received response body response
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:225] Processing body response
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction OUTBOUND: timer disabled
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:386] Continuing processing
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:393] onDestroy
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction INBOUND: timer disabled
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/processor_state.cc:55] Traffic direction OUTBOUND: timer disabled
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/ext_proc.cc:375] Calling close on stream
[2024-07-26 07:21:31.269][17][debug][ext_proc] [source/extensions/filters/http/ext_proc/client_impl.cc:64] Closing gRPC stream
[2024-07-26 07:21:31.270][17][debug][grpc] [source/common/grpc/google_async_client_impl.cc:283] resetStream
[2024-07-26 07:21:31.270][17][debug][grpc] [source/common/grpc/google_async_client_impl.cc:456] Stream cleanup with 2 in-flight tags
[2024-07-26 07:21:31.270][17][debug][grpc] [source/common/grpc/google_async_client_impl.cc:445] Deferred delete
[2024-07-26 07:21:31.271][17][debug][grpc] [source/common/grpc/google_async_client_impl.cc:194] GoogleAsyncStreamImpl destruct
[2024-07-26T07:21:15.534Z] "POST / HTTP/1.1" 200 - 1048576 17 15734 1 "-" "PostmanRuntime/7.36.0" "7db44a81-6622-403d-a69a-3eb5153a265c" "localhost:10000" "192.168.1.202:4000"
[2024-07-26 07:22:05.537][17][debug][grpc] [source/common/grpc/google_async_client_impl.cc:120] Client teardown, resetting streams
```

**Comment by yanjunxiang-google** [comment]
@xr Thanks for testing solution!

**Comment by tyxia** [comment]
> Will also follow the discussions in the PR regarding the [internal buffer](https://github.com/envoyproxy/envoy/pull/34942/files#diff-8ba9034aa20b4759aba1dd6c7a256d33407642c38b0acfbf475bd47382e8ebd8R422).

@xr You probably misunderstood internal buffer in discussion. It is not the one you mentioned above, it is [this one](https://github.com/envoyproxy/envoy/blob/1775547b7ddfa77966442fbae9c5e7ef30a2ccb4/source/extensions/filters/http/ext_proc/processor_state.h#L52) ---> the internal buffer during callout.

 I did highlight it in the internal discussion but it was my bad that I didn't clarify it in the OSS discussion

**Comment by xr** [comment]
closed due to landed in https://github.com/envoyproxy/envoy/pull/34942
